src/ejabberd_auth_pki.erl

```diff
@@ -0,0 +1,117 @@
+-module(ejabberd_auth_pki).
+-author('denys.gonchar@erlang-solutions.com').
+
+-behaviour(ejabberd_gen_auth).
+
+%% ejabberd_gen_auth API
+-export([ start/1,
+          stop/1,
+          store_type/1,
+          set_password/3,
+          authorize/1,
+          try_register/3,
+          dirty_get_registered_users/0,
+          get_vh_registered_users/1,
+          get_vh_registered_users/2,
+          get_vh_registered_users_number/1,
+          get_vh_registered_users_number/2,
+          get_password/2,
+          get_password_s/2,
+          does_user_exist/2,
+          remove_user/2,
+          remove_user/3
+        ]).
+
+-export([ check_password/3,
+          check_password/5]).
+
+%%-callback start(Host :: ejabberd:lserver()) -> ok.
+start(_) -> ok.
+
+%%-callback stop(Host :: ejabberd:lserver()) -> ok.
+stop(_) -> ok.
+
+%%-callback store_type(Host :: ejabberd:lserver()) -> scram | plain | external.
```

Comment:
really scram, or not relevant?
maybe external

Reply:
really, this scram/plain/external store types has nothing common with sasl mechanisms. it's more about the way how we store the password. and we will add certificates support in DB based auth backends later. but that backends must be configured as scram. other way that won't be possible to distinguish between passwords and certificates